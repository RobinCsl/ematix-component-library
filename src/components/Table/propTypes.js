import PropTypes from "prop-types";

export const PropColumn = PropTypes.arrayOf(PropTypes.shape({
  prop: PropTypes.string,
  displayName: PropTypes.string,
  Component: PropTypes.component,
  sortable: PropTypes.bool,
  className: PropTypes.string,
}));


export const PropTypeColumn = PropTypes.shape({
  key: PropTypes.string,
  displayName: PropTypes.string,
  Component: PropTypes.component,
  sortable: PropTypes.bool,
  className: PropTypes.string,
});

