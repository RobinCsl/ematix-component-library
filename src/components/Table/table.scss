.ecl {
  &--table {
    & + * {
      margin-top: 20px;
    }
  }
  &--table-row,
  &--table-header {
    display: flex;
    .ecl--table-cell {
      border-top: 1px solid #aaa;
      flex-grow: 1;
      flex-basis: 100px;
      min-height: 50px;
      display: flex;
      align-items: center;
      padding: 0 20px;
    }
  }

  &--table-header {
    font-weight: 700;
    .ecl--table-cell {
      border-top: none;
    }
  }
}
