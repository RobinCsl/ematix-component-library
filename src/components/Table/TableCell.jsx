import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";

import "./table.scss";

class TableCell extends React.PureComponent {
  render() {
    const {
      className,
      children,
      propsCell,
      propsCell: { className: propClassName },
      ...rest
    } = this.props;

    return (
      <div
        className={classnames("ecl--table-cell", className, propClassName)}
        {...propsCell}
        {...rest}
      >
        {children}
      </div>
    );
  }
}

TableCell.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
  propsCell: PropTypes.object // eslint-disable-line
};

TableCell.defaultProps = {
  propsCell: {},
  className: null,
};

export default TableCell;
