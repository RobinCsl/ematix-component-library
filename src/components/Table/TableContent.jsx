import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";

import "./table.scss";

// components
import TableRow from "./TableRow";

import { PropTypeColumn } from "./propTypes";

class TableContent extends React.PureComponent {
  static propTypes = {
    className: PropTypes.string,
    data: PropTypes.instanceOf(Array).isRequired,
    columns: PropTypes.arrayOf(PropTypeColumn).isRequired,
    uniqueIdentificator: PropTypes.string.isRequired,
    /* eslint-disable react/forbid-prop-types */
    propsContent: PropTypes.object,
    propsRow: PropTypes.object,
    propsRowCell: PropTypes.object,
  };

  static defaultProps = {
    className: null,
    propsContent: {},
    propsRow: {},
    propsRowCell: {},
  };

  componentDidMount() {
    // do sometjing
  }

  render() {
    const {
      className,
      data,
      columns,
      propsContent,
      uniqueIdentificator,
      propsContent: { className: propClassName },
      propsRow,
      propsRowCell,
    } = this.props;

    return (
      <div
        className={classnames(className, propClassName, "ecl--table-content")}
        {...propsContent}
      >
        {data.map(item => (
          <TableRow
            columns={columns}
            //  key is unique (map) so it's fine
            key={item[uniqueIdentificator]} // eslint-disable-line
            item={item}
            propsRow={propsRow}
            propsRowCell={propsRowCell}
          />
        ))}
      </div>
    );
  }
}

export default TableContent;
