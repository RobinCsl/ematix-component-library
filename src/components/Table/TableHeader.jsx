import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";

// coponents
import TableCell from "./TableCell";

// css
import "./table.scss";

import { PropTypeColumn } from "./propTypes";

const TableHeader = ({
  className,
  columns,
  propsHeader,
  propsHeader: { className: propClassName },
  propsHeaderCell,
  ...rest
}) => (
  <div
    className={classnames("ecl--table-header", className, propClassName)}
    {...propsHeader}
    {...rest}
  >
    {columns.map(col => (
      <TableCell propsCell={propsHeaderCell} key={col.key}>
        {col.displayName}
      </TableCell>
    ))}
  </div>
);

TableHeader.propTypes = {
  className: PropTypes.string,
  columns: PropTypes.arrayOf(PropTypeColumn).isRequired,
  /* eslint-disable react/forbid-prop-types */
  propsHeader: PropTypes.object,
  propsHeaderCell: PropTypes.object,
};

TableHeader.defaultProps = {
  className: null,
  propsHeader: {},
  propsHeaderCell: {},
};

export default TableHeader;
