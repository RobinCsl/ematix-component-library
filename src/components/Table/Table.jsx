import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";

import "./table.scss";

import TableContent from "./TableContent";
import TableHeader from "./TableHeader";

// proptypes
import { PropTypeColumn } from "./propTypes";

class Table extends React.PureComponent {
  static propTypes = {
    className: PropTypes.string,
    children: PropTypes.node,
    data: PropTypes.instanceOf(Array).isRequired,
    columns: PropTypes.arrayOf(PropTypeColumn).isRequired,
    uniqueIdentificator: PropTypes.string.isRequired,
    /* eslint-disable react/forbid-prop-types */
    propsContent: PropTypes.object,
    propsRow: PropTypes.object,
    propsRowCell: PropTypes.object,
    propsHeader: PropTypes.object,
    propsHeaderCell: PropTypes.object,
  };

  static defaultProps = {
    className: null,
    children: null,
    propsContent: {},
    propsRow: {},
    propsRowCell: {},
    propsHeader: {},
    propsHeaderCell: {},
  };

  componentDidMount() {
    // do sometjing
  }

  render() {
    const {
      children,
      className,
      data,
      uniqueIdentificator,
      columns,
      propsContent,
      propsRow,
      propsRowCell,
      propsHeader,
      propsHeaderCell,
      ...rest
    } = this.props;

    const tableContentProps = {
      data,
      columns,
      propsContent,
      propsRow,
      propsRowCell,
      uniqueIdentificator,
    };

    const tableHeaderProps = {
      columns,
      propsHeader,
      propsHeaderCell,
    };

    return (
      <div className={classnames(className, "ecl--table")} {...rest}>
        <TableHeader {...tableHeaderProps} />
        <TableContent {...tableContentProps} />
        {children}
      </div>
    );
  }
}

export default Table;
