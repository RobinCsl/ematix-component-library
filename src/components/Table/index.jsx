import Table from "./Table";

export { default as TableCell } from "./TableCell";
export { default as TableRow } from "./TableRow";
export { default as TableContent } from "./TableContent";
export { default as TableHeader } from "./TableHeader";

export default Table;

