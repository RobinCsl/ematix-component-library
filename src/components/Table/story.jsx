import React from "react";
import { storiesOf } from "@storybook/react";
import { Map, Record } from "immutable";
import { action } from "@storybook/addon-actions";

import "./story.scss";

import Codeblock from "../codeblock";
import { Button } from "../index";

// components

import Table from "./Table";
import TableCell from "./TableCell";

// Enum
import Enum from "../../lib/enum";

const howToUse = `
    import {
      Table,
      TableRow,
      TableHeader,
      TableContent,
      TableHeader,
      TableCell
    } from "@ematix/ematix-component-library";
`;
const exampleCode = `
...
`;

const TestRecord = Record({
  id: null,
  name: null,
  status: 0,
});

const EnumStatus = new Enum({
  0: "Active",
  1: "Running",
  2: "Terminated",
});

const plainData = [
  {
    id: "uoaerht",
    name: "Barry Allen",
    status: 1,
  },
  {
    id: "314",
    name: "Harrisson Wells",
    status: 0,
  },
  {
    id: "werthw356",
    name: "Iris West",
    status: 2,
  },
];

const mapItem = item => [item.id, new TestRecord(item)];
const mappedData = new Map(plainData.map(mapItem));

storiesOf("Tables", module).add("Tables", () => (
  <div className="storybook-template">
    <h1 className="title">Tables</h1>
    <p className="description">Description of this cool component</p>
    <h2 className="sub-title">How To Use</h2>
    <div className="codeblock">
      <Codeblock language="javascript" value={howToUse} />
    </div>

    <h2 className="sub-title">UI Props</h2>
    {/* eslint-disable react/prop-types */}
    <Table
      data={mappedData}
      propsRow={{
        onClick: action("row click"),
      }}
      columns={[
        {
          key: "name",
          displayName: "Name",
        },
        {
          key: "status",
          displayName: "Status",
          Component: ({ data: { id, status } }) => (
            <TableCell onClick={action("click cell", id)}>
              {EnumStatus.getBy(status)}
            </TableCell>
          ),
        },
        {
          key: "action",
          displayName: "Status",
          Component: ({ data: { id } }) => (
            <TableCell>
              <Button onClick={action("click button", id)} primary>
                Action
              </Button>
            </TableCell>
          ),
        },
      ]}
    />

    <h2 className="sub-title">Code Example</h2>
    <div className="codeblock">
      <Codeblock language="htmlmixed" value={exampleCode} />
    </div>
  </div>
));
