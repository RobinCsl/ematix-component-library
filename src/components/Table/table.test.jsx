/* eslint-disable */

import React from "react";
import renderer from "react-test-renderer";
import { action } from "@storybook/addon-actions";

import classnames from "classnames";

import "./table.scss";

import Table from "./Table";
import TableCell from "./TableCell";
import Button from "../Button";

// Enum
import Enum from "../../lib/enum";

const EnumStatus = new Enum({
  0: "Active",
  1: "Running",
  2: "Terminated"
});

const plainData = [
  {
    id: "uoaerht",
    name: "Barry Allen",
    status: 1
  },
  {
    id: "314",
    name: "Harrisson Wells",
    status: 0
  },
  {
    id: "werthw356",
    name: "Iris West",
    status: 2
  }
];
const ActionCell = ({ data: { id } }) => (
  <TableCell>
    <Button onClick={action("click button", id)} primary>
      Action
    </Button>
  </TableCell>
);

describe("<Button />", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <div>
          <Table
            data={plainData}
            uniqueIdentificator="id"
            propsRow={{
              onClick: action("row click")
            }}
            columns={[
              {
                key: "name",
                displayName: "Name"
              },
              {
                key: "status",
                displayName: "Status",
                Component: ({ data: { id, status } }) => (
                  <TableCell onClick={action("click cell", id)}>
                    {EnumStatus.getBy(status)}
                  </TableCell>
                )
              },
              {
                key: "action",
                displayName: "Status",
                Component: ActionCell
              }
            ]}
          />
        </div>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
