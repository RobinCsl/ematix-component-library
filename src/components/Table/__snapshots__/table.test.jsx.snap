// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders correctly 1`] = `
<div>
  <div
    className="ecl--table"
  >
    <div
      className="ecl--table-header"
    >
      <div
        className="ecl--table-cell"
      >
        Name
      </div>
      <div
        className="ecl--table-cell"
      >
        Status
      </div>
      <div
        className="ecl--table-cell"
      >
        Status
      </div>
    </div>
    <div
      className="ecl--table-content"
    >
      <div
        className="ecl--table-row"
        onClick={[Function]}
      >
        <div
          className="ecl--table-cell"
        >
          Barry Allen
        </div>
        <div
          className="ecl--table-cell"
          onClick={[Function]}
        >
          Running
        </div>
        <div
          className="ecl--table-cell"
        >
          <button
            className="common-button primary"
            disabled={false}
            onClick={[Function]}
          >
            Action
          </button>
        </div>
      </div>
      <div
        className="ecl--table-row"
        onClick={[Function]}
      >
        <div
          className="ecl--table-cell"
        >
          Harrisson Wells
        </div>
        <div
          className="ecl--table-cell"
          onClick={[Function]}
        >
          Active
        </div>
        <div
          className="ecl--table-cell"
        >
          <button
            className="common-button primary"
            disabled={false}
            onClick={[Function]}
          >
            Action
          </button>
        </div>
      </div>
      <div
        className="ecl--table-row"
        onClick={[Function]}
      >
        <div
          className="ecl--table-cell"
        >
          Iris West
        </div>
        <div
          className="ecl--table-cell"
          onClick={[Function]}
        >
          Terminated
        </div>
        <div
          className="ecl--table-cell"
        >
          <button
            className="common-button primary"
            disabled={false}
            onClick={[Function]}
          >
            Action
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
